sycl/test/virtual-functions/diagnostics-negative.cpp

```diff
@@ -0,0 +1,58 @@
+// RUN: %clangxx -fsycl -fsycl-device-only -c -Xclang -verify %s -o %t.ignored
+//
+// This test is intended to check that correct diagnostics are emitted if
+// kernel attempts to perform virtual calls, but wasn't submitted with the right
+// properties.
+
+#include <sycl/sycl.hpp>
+
+namespace oneapi = sycl::ext::oneapi::experimental;
+
+class Base {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  virtual void foo() const {}
+
+  virtual void bar() const {}
+};
+
+class Derived : public Base {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable<>)
+  void foo() const override {}
+
+  void bar() const override {}
+};
+
+void foo(Base *Ptr) { // expected-note {{performed by function 'foo(Base*)'}}
+  Ptr->foo();
+}
+
+void bar(Base *Ptr) { // expected-note {{performed by function 'bar(Base*)'}}
+  Ptr->bar();
+}
+
+int main() {
+  sycl::queue q;
+
+  Base *Ptr = sycl::malloc_device<Derived>(1, q);
+
+  // expected-note@+1 {{performed by function 'main::'lambda'()::operator()() const'}}
+  q.single_task([=]() { // expected-error {{kernel 'typeinfo name for main::'lambda'()' performs a virtual function call}}
+    Ptr->foo();
```

Comment:
My understanding is that this fails because it doesn't have `oneapi::calls_indirectly` property. Should we also check failures when we do have `oneapi::calls_indirectly` but the function doesn't have `oneapi::indirectly_callable`?

Reply:
> My understanding is that this fails because it doesn't have `oneapi::calls_indirectly` property. Should we also check failures when we do have `oneapi::calls_indirectly` but the function doesn't have `oneapi::indirectly_callable`?

We won't be able to detect that, LLVM IR at call site looks like:
```
%vtable = load [some type] gep %this, [some indices]
%ptr = gep ptr %vtable, [some index]
call %ptr([some args]) # "virtual-call"
```

We have no information about which exact function is called here and therefore whether or not it has a property attached to it.